misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java

```diff
@@ -244,15 +290,90 @@ static Token formatToken(Token token, int leadingSpaces, int trailingSpaces, int
         if (token == null) {
             return token;
         }
-        MinutiaeList newLeadingMinutiaeList = modifyMinutiaeList(leadingSpaces, leadingNewLines);
-        MinutiaeList newTrailingMinutiaeList = modifyMinutiaeList(trailingSpaces, trailingNewLines);
+        MinutiaeList newLeadingMinutiaeList = preserveComments(token.leadingMinutiae(), trailingNewLines)
+                .add(createWhitespaceMinutiae(getWhiteSpaces(leadingSpaces, leadingNewLines)));
+        MinutiaeList newTrailingMinutiaeList = preserveComments(token.trailingMinutiae(), trailingNewLines)
+                .add(createWhitespaceMinutiae(getWhiteSpaces(trailingSpaces, trailingNewLines)));
 
         return token.modify(newLeadingMinutiaeList, newTrailingMinutiaeList);
     }
 
-    private static MinutiaeList modifyMinutiaeList(int spaces, int newLines) {
-        Minutiae minutiae = createWhitespaceMinutiae(getWhiteSpaces(spaces, newLines));
-        return createMinutiaeList(minutiae);
+    private static MinutiaeList preserveComments(MinutiaeList minutiaeList, int trailingNewLines) {
+        MinutiaeList minutiaes = AbstractNodeFactory.createEmptyMinutiaeList();
+        if (minutiaeList.size() > 0) {
+            int count = commentCount(minutiaeList);
+            if (count > 0) {
+                int processedCount = 0;
+                for (int i = 0; i < minutiaeList.size(); i++) {
+                    Minutiae minutiae = minutiaeList.get(i);
+                    minutiaes = minutiaes.add(minutiae);
+                    if (minutiae.kind() == SyntaxKind.COMMENT_MINUTIAE) {
+                        processedCount++;
+                        if (processedCount == count) {
+                            if (trailingNewLines == 0) {
+                                minutiaes = minutiaes.add(AbstractNodeFactory
+                                        .createEndOfLineMinutiae(System.getProperty(LINE_SEPARATOR)));
+                            }
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        return minutiaes;
+    }
+
+    private static int commentCount(MinutiaeList minutiaeList) {
+        int count = 0;
+        for (int i = 0; i < minutiaeList.size(); i++) {
+            if (minutiaeList.get(i).kind() == SyntaxKind.COMMENT_MINUTIAE) {
```

Comment:
Shall we invert the condition with a continue?

Reply:
This method checks the number of comment minutiaes available inside the minutiae list. Therefore, the entire list should be looped.